### PR TITLE
flowctl: raw suggest-schema command

### PR DESCRIPTION
**Pull request by mdibaiee**
**Description:**

reads a collection and ops logs of the relevant task, and runs schema inference on documents of the collection and documents that have violated schema validation to come up with a schema that will allow those documents to pass through the task. runs `git diff` on the resulting schema and the original schema to give an idea of how a patch may be constructed.

**Workflow steps:**

```
flowctl raw suggest-schema
  --collection estuary/marketing/google-ads/campaigns
  --task estuary/marketing/google-ads/source-google-ads
```

**Documentation links affected:**

(list any [documentation links](https://docs.google.com/document/d/1SRC9VS9zyCzWl3n4HXHbc4wPB1eLxJHkA2rtu9ZNokM/edit?usp=sharing) that you created, or existing ones that you've identified as needing updates, along with a brief description)

**Notes for reviewers:**

(anything that might help someone review this PR)

- - -
This change is [Reviewable](https://reviewable.io/reviews/estuary/flow/1093)

### CODE REVIEW

**Review by psFried — CHANGES_REQUESTED** [review]
I had a bunch of questions and comments here, but a lot of them are probably just due to me not understanding all the context around how this fits into the broader workflow. I'd mentioned a few specific areas, and I'll repeat here, that comments could have really helped me understand this better. The code is self-documenting of _what_ (although in some cases having a summary is still helpful), but I would especially appreciate some documentation of _why_ things are the way that they are.

I'm also still not clear on why this reads both the collection documents and the ops logs. At first blush, it seems like reading the collection documents would be unnecessary, since the documents there have clearly already validated against the collection schema. If we do need to read the collection, does it really have to be the _full_ collection, or can we support specifying a `ReadBounds` to limit the amount of data we have to fetch?

File: crates/flowctl/src/raw/suggest_schema.rs
Comment (mdibaiee) [reply crates/flowctl/src/raw/suggest_schema.rs:59]: @psFried the default for `SpecTypeSelector` fields is `None`, and `None` behaves the same as `Some(true)`, see [the comments there](https://github.com/estuary/flow/blob/master/crates/flowctl/src/catalog/mod.rs#L122-L127):

````
/// Whether to include captures in the selection
    ///
    /// If true, or if no value was given, then captures will
    /// be included. You can also use `--captures=false` to exclude
    /// captures.
```

Also see [this logic](https://github.com/estuary/flow/blob/master/crates/flowctl/src/catalog/mod.rs#L163C63-L170) which explicitly unwraps the `Option` value with a default value of `false`.

Comment (mdibaiee) [reply crates/flowctl/src/raw/suggest_schema.rs:68]: Actually I'm confused about the second logic that I linked above, if the `Option` is unwrapped to `false` and the spec type is filtered out, does that mean we just don't specify any query for that spec type, and does that mean it will be included by default?

File: crates/flowctl/src/raw/suggest_schema.rs
Comment (mdibaiee) [reply crates/flowctl/src/raw/suggest_schema.rs]: The original purpose of the files is to run diff, but I personally find it useful to be able to see the actual files if I'm using this command, to be able to better understand what is the full resulting schema, and sometimes compare it manually to the existing schema. It also sometimes makes it easier to copy the new inferred schema from this file rather than from the diff (which is mixed with `+` and `-` and other signs in between).

Regarding the naming of the files... we could name them after the collection to make it easier to keep them organised and not overwrite files frequently

File: crates/flowctl/src/raw/suggest_schema.rs
Comment (mdibaiee) [reply crates/flowctl/src/raw/suggest_schema.rs:55]: It can be either a materialization or a capture, but I think most of the time a capture will be used. I've added some comments to this command, let me know if you think this could still be improved

File: crates/flowctl/src/raw/suggest_schema.rs
Comment (psFried) [reply crates/flowctl/src/raw/suggest_schema.rs:75]: The logic for this is [here](https://github.com/estuary/flow/blob/58b6586fa160340955c0ea6806b9676b8174d654/crates/flowctl/src/catalog/mod.rs#L175-L197). If any `SpecTypeSelector` fields are `Some(true)`, then it'll query for specs having only that spec type. If none are `true` and some are `Some(false)`, then it will only filter out the types that you don't want. If all fields are `None`, then it just returns everything.

**Review by psFried — CHANGES_REQUESTED** [review]
We caught on VC, and said that we're still working on figuring out the overall workflow for getting suggested schemas into connectors. Once we have a better idea of that, we'll make a README with some instructions. Given that, I'd like to suggest that we hold off on merging this PR until that guide is complete. That's assuming that there isn't any reason to merge earlier, so LMK if there is one.

File: crates/flowctl/src/raw/suggest_schema.rs
Comment (mdibaiee) [reply crates/flowctl/src/raw/suggest_schema.rs:36]: @psFried the documents will have passed validation but that's because these collections all have `additionalProperties: true`, so missing fields are prominent among them.

File: crates/flowctl/src/raw/suggest_schema.rs
Comment (mdibaiee) [reply crates/flowctl/src/raw/suggest_schema.rs:173]: `diff -u` does not give me colored output unfortunately, even when I explicitly ask it to using `diff -u --color=always`, in that sense the `git diff` is still better for readability.

This is informative at this stage, and aims to help the person using it write a patch file themselves. Hopefully at some point we can produce a proper JSON Merge Patch automatically (I couldn't find a crate that does this, so we'll have to write that at some point, shouldn't be complex).

**Comment by mdibaiee** [comment]
> Given that, I'd like to suggest that we hold off on merging this PR until that guide is complete. That's assuming that there isn't any reason to merge earlier, so LMK if there is one.

I would like to merge so Dave and others can download the latest binary and use it as well without having to manually build

File: crates/flowctl/src/raw/suggest_schema.rs
Comment (psFried) [reply crates/flowctl/src/raw/suggest_schema.rs:177]: Thanks, that all makes sense. I do think it'd be good to explicitly call out in the comments that this diff is informational, whereas the file outputs are (presumably) what the user actually wants out of this.

**Review by psFried — CHANGES_REQUESTED** [review]
> I would like to merge so Dave and others can download the latest binary and use it as well without having to manually build.

Sounds good to me. I left just a few more comments, but we should be good to merge after those are addressed.

File: crates/flowctl/src/raw/suggest_schema.rs
Comment (mdibaiee) [reply crates/flowctl/src/raw/suggest_schema.rs:141]: When you say lossy, you mean in terms of the original JSON structure of the file, right? Not in terms of the meaning of the jsonschema.

The reason for this is that we want to have two JSONSchema structures that are similar in layout, which would allow us to run a more intuitive diff, and as long as the meaning of the JSONSchema stays the same, this diff would be informative for the user to be able to write a JSON Merge Patch.

File: crates/flowctl/src/raw/suggest_schema.rs
Comment (mdibaiee) [reply crates/flowctl/src/raw/suggest_schema.rs:36]: With `ReadArgs`, by setting `since=now` you can skip this at the moment

**Comment by mdibaiee** [comment]
Docs PR: https://github.com/estuary/airbyte/pull/211

**Review by psFried — APPROVED** [review]
LGTM

**Pull request merged**
